File: spark/src/main/scala/com/examples/hops/spark/stark/PolygonsPolygons.scala
package com.examples.hops.spark.stark

import java.util.logging.Logger

import com.examples.hops.spark.utils.HDFSWriter
import dbis.stark.STObject
import dbis.stark.spatial.JoinPredicate
import dbis.stark.spatial.partitioner.SpatialGridPartitioner
import org.apache.spark.sql.SparkSession

object PolygonsPolygons {

  val log: Logger = Logger.getLogger("PolygonsPolygons")

  def main(args: Array[String]): Unit = {

    var n_partitions = 1
    var n_cores = 8

    if (args.length > 0) {
      n_partitions = Integer.parseInt(args(0))
      if (args.length > 1)
        n_cores = Integer.parseInt(args(1))
    }

    val comments = "Polygons Contains Polygons"
    val path = "hdfs:///Projects/demo_spark_kgiann01/Resources/PolygonsContainPolygons_p" + n_partitions + "_c" + n_cores + ".txt"

    val spark = SparkSession
      .builder
      .appName("PolygonsContainsPolygons")
      .getOrCreate

    val sc = spark.sparkContext

    /* read data */
    val arealm = sc.textFile("hdfs:///Projects/demo_spark_kgiann01/Resources/arealm_merge.tsv")
      .map(line => line.split('\t'))
      .map(arr =>
        (STObject(arr(0)), arr(6))
      )


    var polygonsPolygonExecutor: PolygonsPolygonExecutor = null;

    if (n_partitions > 1) {
      /* fixed grid partitioner */
      val arealmGridPartitioner = new SpatialGridPartitioner(arealm, partitionsPerDimension = n_partitions, false, 2)

      polygonsPolygonExecutor = new PolygonsPolygonExecutor(arealmGridPartitioner, arealm)
    } else {
      polygonsPolygonExecutor = new PolygonsPolygonExecutor(null, arealm)
    }

    /* warmup run */
    var start = System.currentTimeMillis()
    var result = polygonsPolygonExecutor.getResult(JoinPredicate.CONTAINS)
    var count = result.toJavaRDD().count()
    val warmupDuration = System.currentTimeMillis() - start

    val iterations = 3
    val i = 0
    /* actual runs */
    var totalDuration: Long = 0
    for (i <- 1 to iterations) {
      start = System.currentTimeMillis()
      result = polygonsPolygonExecutor.getResult(JoinPredicate.CONTAINS)
      count = result.toJavaRDD().count()
      totalDuration += System.currentTimeMillis() - start
    }

    val avgDuration = totalDuration / iterations

    val writer = new HDFSWriter(path)
    writer.write(comments, false, count, avgDuration.toDouble / 1000, warmupDuration.toDouble / 1000, iterations, n_partitions, n_cores)
    writer.close()
  }
}

File: spark/geospark/src/main/java/com/examples/hops/spark/geospark/PolygonsPolygons.java
package com.examples.hops.spark.geospark;

import com.vividsolutions.jts.geom.Polygon;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.storage.StorageLevel;
import org.datasyslab.geospark.enums.FileDataSplitter;
import org.datasyslab.geospark.enums.GridType;
import org.datasyslab.geospark.spatialOperator.JoinQuery;
import org.datasyslab.geospark.spatialRDD.PolygonRDD;

import java.io.IOException;
import java.util.HashSet;
import java.util.logging.Logger;

public class PolygonsPolygons {

    private static final Logger log = Logger.getLogger(PolygonsPolygons.class.getName());

    public static void main(String[] args) throws Exception {
        SparkSession spark = SparkSession
                .builder()
                .appName("GeoSpark_PolygonsContainsPolygons")
                .getOrCreate();

        JavaSparkContext sc = new JavaSparkContext(spark.sparkContext());
        String arealm = "hdfs:///Projects/demo_spark_kgiann01/Resources/arealm_merge.txt";
        FileDataSplitter splitter = FileDataSplitter.WKT; // input file format
        Integer n_partitions = 1; // number of partitions
        Integer n_cores = 8; // number of partitions

        long resultSetCount;
        long tStartRDD = System.currentTimeMillis();
        //create PointRDD from datasets NO INDEX
        PolygonRDD objectPolygonARDD = new PolygonRDD(sc, arealm, splitter, true, n_partitions, StorageLevel.MEMORY_ONLY());
        objectPolygonARDD.spatialPartitioning(GridType.EQUALGRID);
        objectPolygonARDD.rawSpatialRDD.persist(StorageLevel.MEMORY_ONLY());
        PolygonRDD objectPolygonBRDD = new PolygonRDD(sc, arealm, splitter, true, n_partitions, StorageLevel.MEMORY_ONLY());
        objectPolygonBRDD.spatialPartitioning(objectPolygonARDD.grids);
        objectPolygonBRDD.rawSpatialRDD.persist(StorageLevel.MEMORY_ONLY());
        long tEndRDD = System.currentTimeMillis();

        long tDelta = tEndRDD - tStartRDD;
        double elapsedSecondsRDD = tDelta / 1000.0;
        double elapsedSeconds = 0.0;

        long tStart = System.currentTimeMillis();
        JavaPairRDD<Polygon, HashSet<Polygon>> results = JoinQuery.SpatialJoinQuery(objectPolygonARDD, objectPolygonBRDD, false, true); //first execution to update the cache
        resultSetCount = results.count();
        long tEnd = System.currentTimeMillis();

        tDelta = tEnd - tStart;

        double curElapsedSeconds = tDelta / 1000.0;
        double warmupTime = curElapsedSeconds;
        int iter = 3;

        for (int i = 0; i < iter; i++) {

            tStart = System.currentTimeMillis();
            results = JoinQuery.SpatialJoinQuery(objectPolygonARDD, objectPolygonBRDD, false, true);
            resultSetCount = results.count();
            tEnd = System.currentTimeMillis();
            tDelta = tEnd - tStart;
            curElapsedSeconds = tDelta / 1000.0;
            elapsedSeconds = curElapsedSeconds + elapsedSeconds;

        }

        double avgExecTime = elapsedSeconds / iter;
        try (HDFSWriter writer = new HDFSWriter("hdfs:///Projects/demo_spark_kgiann01/Resources/GeosparkPolygonsPolygonsNoIndex"+n_partitions+".txt")) {
            writer.write(PolygonsPolygons.class.getName(), false, resultSetCount, avgExecTime, warmupTime, iter, n_partitions, n_cores);
        } catch (IOException e) {
            e.printStackTrace();
        }
        spark.stop();
    }
}
